examples/bq_file_load_benchmark/benchmark_tools/staging_table_generator.py

```diff
@@ -0,0 +1,221 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import logging
+import os
+
+from google.cloud import bigquery
+
+
+class StagingTableGenerator(object):
+    """Contains methods for generating staging tables in BigQuery.
+
+    Utilizes the Dataflow Data Generator tool from the PSO library. Uses the
+    dataflow_data_generator module to create staging tables using the columnType
+    and numColumn parameters, and then the bq_table_resizer module to create r
+    resized staging tables with the targetDataSizes parameter.
+
+    Attributes:
+         project(str): ID of the project that holds the staging tables and
+            resized staging tables.
+         staging_dataset_id(str): ID of the dataset that holds the staging
+            tables.
+         resized_dataset_id(str): ID of the dataset that holds the
+            resized staging tables.
+         json_schema_path(str): Directory that holds the json schemas used
+            to create the staging tables.
+
+    """
+
+    def __init__(
+            self,
+            project,
+            staging_dataset_id,
+            resized_dataset_id,
+            json_schema_path,
+            file_params,
+            num_rows
+    ):
+        self.bq_client = bigquery.Client()
+        self.project = project
+        self.staging_dataset_id = staging_dataset_id
+        self.resized_dataset_id = resized_dataset_id
+        self.json_schema_path = json_schema_path
+        self.file_params = file_params
+        self.num_rows = num_rows
+
+    def create_staging_tables(
+            self,
+            dataflow_staging_location,
+            dataflow_temp_location,
+    ):
+        """Creates staging tables using the columnType and numColumn parameters.
+
+        Utilizes the data_generator_pipeline module from the Dataflow Data
+        Generator tool to create staging tables. Names of schema combinations
+        are created using the columnTypes and numColumn parameters. Then the
+        schema names are used to obtain json schemas from the provided
+        self.json_schema_path, which are then fed into to data generator
+        pipeline to create staging tables.
+
+        Args:
+            dataflow_staging_location(str): GCS staging path.
+            dataflow_temp_location(str): GCS temp path.
+        """
+        column_types = self.file_params['columnTypes']
+        num_columns = self.file_params['numColumns']
+        abs_path = os.path.abspath(os.path.dirname(__file__))
+        data_gen_path = os.path.join(
+            abs_path,
+            '../../dataflow-data-generator/data-generator-pipeline'
+        )
+
+        schema_names = []
+        for column_type in column_types:
+            for num_column in num_columns:
+                schema_names.append('{0:s}_{1:d}'.format(
+                    column_type,
+                    num_column
+                ))
+        for schema_name in schema_names:
```

Comment:
You could kick these jobs off in parallel if project has sufficient resources.

Reply:
I tried the following: `    def create_staging_tables(
            self,
            dataflow_staging_location,
            dataflow_temp_location,
    ):
        """Creates staging tables using the columnType and numColumn parameters.

        Utilizes the data_generator_pipeline module from the Dataflow Data
        Generator tool to create staging tables. Names of schema combinations
        are created using the columnTypes and numColumn parameters. Then the
        schema names are used to obtain json schemas from the provided
        self.json_schema_path, which are then fed into to data generator
        pipeline to create staging tables.

        Args:
            dataflow_staging_location(str): GCS staging path.
            dataflow_temp_location(str): GCS temp path.
        """

        def _create_table(table_details):

            column_type, num_column = table_details
            schema_name = '{0:s}_{1:d}'.format(
                column_type,
                num_column
            )
            logging.info('Creating staging table for schema: {0:s}'.format(
                schema_name
            ))
            command_str = ('python {0:s}/data_generator_pipeline.py '
                           '--schema_file={1:s}/{2:s}.json '
                           '--num_records={3:d} '
                           '--output_bq_table={4:s}:{5:s}.{2:s} '
                           '--project={4:s} '
                           '--setup_file={0:s}/setup.py '
                           '--staging_location={6:s} '
                           '--temp_location={7:s} '
                           '--save_main_session '
                           '--worker_machine_type=n1-highcpu-32 '
                           '--runner=DataflowRunner ')
            command = command_str.format(
                data_gen_path,
                self.json_schema_path,
                schema_name,
                self.num_rows,
                self.project,
                self.staging_dataset_id,
                dataflow_staging_location,
                dataflow_temp_location,
            )
            os.system(command)

        column_types = self.file_params['columnTypes']
        num_columns = self.file_params['numColumns']
        abs_path = os.path.abspath(os.path.dirname(__file__))
        data_gen_path = os.path.join(
            abs_path,
            '../../dataflow-data-generator/data-generator-pipeline'
        )
        with ThreadPoolExecutor() as p:
            p.map(_create_table, itertools.product(column_types, num_columns))`

It looks like it does kick of _create_table in parallel, because I immediately got this output from the logging.info in _create_table: 
`INFO:root:Creating staging table for schema: 100_STRING_10
INFO:root:Creating staging table for schema: 100_STRING_100
INFO:root:Creating staging table for schema: 100_STRING_1000
INFO:root:Creating staging table for schema: 50_STRING_50_NUMERIC_10
INFO:root:Creating staging table for schema: 50_STRING_50_NUMERIC_100
INFO:root:Creating staging table for schema: 50_STRING_50_NUMERIC_1000
INFO:root:Creating staging table for schema: 10_STRING_90_NUMERIC_10
INFO:root:Creating staging table for schema: 10_STRING_90_NUMERIC_100
INFO:root:Creating staging table for schema: 10_STRING_90_NUMERIC_1000`

But when I look at the dataflow console, I only see one dataflow job running for me. So I'm guessing the other threads are waiting to start dataflow until the current job is done?